flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/metrics/SubStateTimeMetricsTest.java

```diff
@@ -0,0 +1,624 @@
+                        JobType.BATCH, settings, ExecutionState.INITIALIZING, "initialising");
+
+        final ExecutionAttemptID id1 = createExecutionAttemptId();
+
+        metrics.onStateUpdate(id1, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+        metrics.onStateUpdate(id1, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+        metrics.onStateUpdate(id1, ExecutionState.DEPLOYING, ExecutionState.INITIALIZING);
+        metrics.onStateUpdate(id1, ExecutionState.INITIALIZING, ExecutionState.RUNNING);
+
+        final ExecutionAttemptID id2 = createExecutionAttemptId();
+
+        metrics.onStateUpdate(id2, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+        metrics.onStateUpdate(id2, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+        metrics.onStateUpdate(id2, ExecutionState.DEPLOYING, ExecutionState.INITIALIZING);
+
+        assertThat(metrics.getBinary()).isEqualTo(0L);
+    }
+
+    @Test
+    void testOneInitializingAndOneDeploying_batch() {
+        final SubStateTimeMetrics metrics =
+                new SubStateTimeMetrics(
+                        JobType.BATCH, settings, ExecutionState.INITIALIZING, "initialising");
+
+        final ExecutionAttemptID id1 = createExecutionAttemptId();
+        final ExecutionAttemptID id2 = createExecutionAttemptId();
+
+        metrics.onStateUpdate(id1, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+        metrics.onStateUpdate(id2, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+
+        metrics.onStateUpdate(id1, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+        metrics.onStateUpdate(id2, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+
+        metrics.onStateUpdate(id1, ExecutionState.DEPLOYING, ExecutionState.INITIALIZING);
+        assertThat(metrics.getBinary()).isEqualTo(1L);
+    }
+
+    @Test
+    void testOneInitializingAndOneDeploying_streaming() {
+        final SubStateTimeMetrics metrics =
+                new SubStateTimeMetrics(
+                        JobType.STREAMING, settings, ExecutionState.INITIALIZING, "initialising");
+
+        final ExecutionAttemptID id1 = createExecutionAttemptId();
+        final ExecutionAttemptID id2 = createExecutionAttemptId();
+
+        metrics.onStateUpdate(id1, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+        metrics.onStateUpdate(id2, ExecutionState.CREATED, ExecutionState.SCHEDULED);
+
+        metrics.onStateUpdate(id1, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+        metrics.onStateUpdate(id2, ExecutionState.SCHEDULED, ExecutionState.DEPLOYING);
+
+        metrics.onStateUpdate(id1, ExecutionState.DEPLOYING, ExecutionState.INITIALIZING);
+        assertThat(metrics.getBinary()).isEqualTo(0L);
+
+        metrics.onStateUpdate(id2, ExecutionState.DEPLOYING, ExecutionState.INITIALIZING);
+        assertThat(metrics.getBinary()).isEqualTo(1L);
+    }
+
+    @Test
+    void testInitialzingEnd_streaming_ignoresTerminalDeployments() {
```

Comment:
I think the test is the only place where we test specific transitions for each supported metric -- and this is the only place that will hold the specifics for future metrics that we add.

The reason as to why this test is valuable is because it allows us to check metric level correctness, and also exercises the entire chain (SubStateTimeMetrics, MetricsProvider).

The duplication is because the state transitions and expected values for the given metrics are different.

Unless you have a strong objection to this, I would rather retain this test.